File: frontend/src/lib/api/conflicts.ts
import { apiRequest } from "@/lib/api/request";

const BASE = "/admin/permissions";

// ── Types ─────────────────────────────────────────────────────────────────────

export type ConflictSeverity = "info" | "warning" | "blocking" | "security_risk";
export type ConflictStatus = "open" | "investigating" | "resolved" | "dismissed";

export type ConflictEntry = {
  id: string;
  organization_id: string;
  subject_type: string;
  subject_value: string;
  user_id: string | null;
  role_name: string | null;
  resource_type: string;
  resource_id: string | null;
  action: string;
  conflict_type: string;
  severity: ConflictSeverity;
  status: ConflictStatus;
  detected_at: string;
  resolved_at: string | null;
  conflict_summary: string | null;
  grant_id: string | null;
  deny_id: string | null;
  remediation: string[];
  context: Record<string, unknown>;
};

export type ConflictListResponse = {
  items: ConflictEntry[];
  total: number;
  page: number;
  page_size: number;
};

export type UpdateConflictStatusRequest = {
  status: "investigating" | "resolved" | "dismissed";
  resolution_note?: string | null;
};

export type ScanResult = {
  conflicts_detected: number;
  conflicts_created: number;
  scan_duration_ms: number;
  scanned_grants: number;
  scanned_denies: number;
  scanned_acl_mappings: number;
};

export type TraceStep = {
  rule: string;
  outcome: "pass" | "allow" | "deny";
  detail: string | null;
};

export type ExplainDecisionResponse = {
  decision: "allow" | "deny";
  matched_rule: string;
  deny_reason: string | null;
  subject_user_id: string;
  resource_type: string;
  resource_id: string | null;
  action: string;
  trace: TraceStep[];
  remediation: string[];
  request_id: string;
};

// ── Normalizers ────────────────────────────────────────────────────────────────

function normalizeConflictEntry(value: unknown): ConflictEntry {
  const raw = value && typeof value === "object" ? (value as Record<string, unknown>) : {};
  return {
    id: typeof raw.id === "string" ? raw.id : "",
    organization_id: typeof raw.organization_id === "string" ? raw.organization_id : "",
    subject_type: typeof raw.subject_type === "string" ? raw.subject_type : "",
    subject_value: typeof raw.subject_value === "string" ? raw.subject_value : "",
    user_id: typeof raw.user_id === "string" ? raw.user_id : null,
    role_name: typeof raw.role_name === "string" ? raw.role_name : null,
    resource_type: typeof raw.resource_type === "string" ? raw.resource_type : "",
    resource_id: typeof raw.resource_id === "string" ? raw.resource_id : null,
    action: typeof raw.action === "string" ? raw.action : "",
    conflict_type: typeof raw.conflict_type === "string" ? raw.conflict_type : "",
    severity: (raw.severity as ConflictSeverity) ?? "info",
    status: (raw.status as ConflictStatus) ?? "open",
    detected_at: typeof raw.detected_at === "string" ? raw.detected_at : "",
    resolved_at: typeof raw.resolved_at === "string" ? raw.resolved_at : null,
    conflict_summary: typeof raw.conflict_summary === "string" ? raw.conflict_summary : null,
    grant_id: typeof raw.grant_id === "string" ? raw.grant_id : null,
    deny_id: typeof raw.deny_id === "string" ? raw.deny_id : null,
    remediation: Array.isArray(raw.remediation)
      ? (raw.remediation as unknown[]).filter((r): r is string => typeof r === "string")
      : [],
    context:
      raw.context && typeof raw.context === "object"
        ? (raw.context as Record<string, unknown>)
        : {},
  };
}

function normalizeTraceStep(value: unknown): TraceStep {
  const raw = value && typeof value === "object" ? (value as Record<string, unknown>) : {};
  return {
    rule: typeof raw.rule === "string" ? raw.rule : "",
    outcome: (raw.outcome as TraceStep["outcome"]) ?? "pass",
    detail: typeof raw.detail === "string" ? raw.detail : null,
  };
}

// ── API functions ──────────────────────────────────────────────────────────────

export async function listConflicts(params?: {
  severity?: ConflictSeverity;
  status?: ConflictStatus;
  resource_type?: string;
  page?: number;
  page_size?: number;
}): Promise<ConflictListResponse> {
  const qs = new URLSearchParams();
  if (params?.severity) qs.set("severity", params.severity);
  if (params?.status) qs.set("status", params.status);
  if (params?.resource_type) qs.set("resource_type", params.resource_type);
  if (params?.page) qs.set("page", String(params.page));
  if (params?.page_size) qs.set("page_size", String(params.page_size));
  const url = `${BASE}/conflicts${qs.size ? `?${qs}` : ""}`;
  const payload = await apiRequest<unknown>(url, { method: "GET", retry: false });
  const raw = payload && typeof payload === "object" ? (payload as Record<string, unknown>) : {};
  return {
    items: Array.isArray(raw.items) ? raw.items.map(normalizeConflictEntry) : [],
    total: typeof raw.total === "number" ? raw.total : 0,
    page: typeof raw.page === "number" ? raw.page : 1,
    page_size: typeof raw.page_size === "number" ? raw.page_size : 50,
  };
}

export async function getConflict(conflictId: string): Promise<ConflictEntry> {
  const payload = await apiRequest<unknown>(
    `${BASE}/conflicts/${encodeURIComponent(conflictId)}`,
    { method: "GET", retry: false },
  );
  return normalizeConflictEntry(payload);
}

export async function updateConflictStatus(
  conflictId: string,
  req: UpdateConflictStatusRequest,
): Promise<ConflictEntry> {
  const payload = await apiRequest<unknown>(
    `${BASE}/conflicts/${encodeURIComponent(conflictId)}/status`,
    { method: "PATCH", json: req, retry: false },
  );
  return normalizeConflictEntry(payload);
}

export async function scanForConflicts(): Promise<ScanResult> {
  const payload = await apiRequest<unknown>(`${BASE}/conflicts/scan`, {
    method: "POST",
    retry: false,
  });
  const raw = payload && typeof payload === "object" ? (payload as Record<string, unknown>) : {};
  return {
    conflicts_detected: typeof raw.conflicts_detected === "number" ? raw.conflicts_detected : 0,
    conflicts_created: typeof raw.conflicts_created === "number" ? raw.conflicts_created : 0,
    scan_duration_ms: typeof raw.scan_duration_ms === "number" ? raw.scan_duration_ms : 0,
    scanned_grants: typeof raw.scanned_grants === "number" ? raw.scanned_grants : 0,
    scanned_denies: typeof raw.scanned_denies === "number" ? raw.scanned_denies : 0,
    scanned_acl_mappings: typeof raw.scanned_acl_mappings === "number" ? raw.scanned_acl_mappings : 0,
  };
}

export async function explainDecision(params: {
  subject_user_id: string;
  resource_type: string;
  action: string;
  resource_id?: string | null;
}): Promise<ExplainDecisionResponse> {
  const qs = new URLSearchParams({
    subject_user_id: params.subject_user_id,
    resource_type: params.resource_type,
    action: params.action,
  });
  if (params.resource_id) qs.set("resource_id", params.resource_id);
  const payload = await apiRequest<unknown>(
    `${BASE}/explain-decision?${qs}`,
    { method: "GET", retry: false },
  );
  const raw = payload && typeof payload === "object" ? (payload as Record<string, unknown>) : {};
  return {
    decision: raw.decision === "allow" ? "allow" : "deny",
    matched_rule: typeof raw.matched_rule === "string" ? raw.matched_rule : "",
    deny_reason: typeof raw.deny_reason === "string" ? raw.deny_reason : null,
    subject_user_id: typeof raw.subject_user_id === "string" ? raw.subject_user_id : "",
    resource_type: typeof raw.resource_type === "string" ? raw.resource_type : "",
    resource_id: typeof raw.resource_id === "string" ? raw.resource_id : null,
    action: typeof raw.action === "string" ? raw.action : "",
    trace: Array.isArray(raw.trace) ? raw.trace.map(normalizeTraceStep) : [],
    remediation: Array.isArray(raw.remediation)
      ? (raw.remediation as unknown[]).filter((r): r is string => typeof r === "string")
      : [],
    request_id: typeof raw.request_id === "string" ? raw.request_id : "",
  };
}

File: backend/app/domains/permissions/schemas/conflicts.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

# ── Conflict severity / status constants ──────────────────────────────────────

# DB uses: low / medium / high / critical
# API surface uses: info / warning / blocking / security_risk (maps below)
SEVERITY_TO_DB: dict[str, str] = {
    "info": "low",
    "warning": "medium",
    "blocking": "high",
    "security_risk": "critical",
}
DB_TO_SEVERITY: dict[str, str] = {v: k for k, v in SEVERITY_TO_DB.items()}

CONFLICT_TYPES = (
    "role_allow_resource_deny",
    "collection_allow_connector_acl_deny",
    "citation_visible_source_hidden",
    "graph_entity_visible_evidence_inaccessible",
    "stale_grant_deleted_resource",
    "stale_grant_removed_connector",
    "orphaned_acl_mapping",
    "feature_deny_active_grant",
    "explicit_grant_conflicts_role_deny",
)

CONFLICT_STATUSES = ("open", "investigating", "resolved", "dismissed")

SEVERITY_LABELS = ("info", "warning", "blocking", "security_risk")

_REMEDIATION: dict[str, list[str]] = {
    "role_allow_resource_deny": [
        "Review the explicit deny entry and remove it if access should be granted.",
        "If the deny is intentional, revoke the conflicting grant.",
        "Consider whether this principal requires a narrower role instead of a broad grant.",
    ],
    "collection_allow_connector_acl_deny": [
        "Re-sync the connector ACL to ensure collection-level access is reflected.",
        "Remove the collection grant if the connector ACL restriction is correct.",
        "Contact the connector administrator to update ACL permissions upstream.",
    ],
    "stale_grant_deleted_resource": [
        "Revoke the grant as the target resource no longer exists.",
        "Audit other grants from the same principal for additional stale entries.",
    ],
    "stale_grant_removed_connector": [
        "Revoke the connector grant and re-create it if the connector is re-connected.",
        "Verify the connector is still active before granting connector-scoped access.",
    ],
    "orphaned_acl_mapping": [
        "Remove ACL mappings for connectors that have been deleted or disconnected.",
        "Re-run the connector sync to generate fresh ACL mappings.",
    ],
    "feature_deny_active_grant": [
        "If the feature is intentionally disabled, revoke conflicting explicit grants.",
        "Enable the feature for this organisation if grant-level access is correct.",
    ],
    "explicit_grant_conflicts_role_deny": [
        "Review whether the explicit grant is intentional given the role restriction.",
        "Downgrade the principal's role if the grant should be the limiting factor.",
    ],
    "citation_visible_source_hidden": [
        "Revoke citation-level access until the underlying source is also accessible.",
        "Grant the principal access to the source document backing the citation.",
    ],
    "graph_entity_visible_evidence_inaccessible": [
        "Ensure the principal has access to evidence documents backing the entity.",
        "If evidence documents are restricted, restrict graph entity access to match.",
    ],
}


def remediation_for(conflict_type: str) -> list[str]:
    return _REMEDIATION.get(conflict_type, ["Review this conflict manually with an administrator."])


# ── Response models ────────────────────────────────────────────────────────────


class ConflictResponse(BaseModel):
    id: str
    organization_id: str
    subject_type: str
    subject_value: str
    user_id: str | None
    role_name: str | None
    resource_type: str
    resource_id: str | None
    action: str
    conflict_type: str
    severity: str  # API surface: info/warning/blocking/security_risk
    status: str
    detected_at: datetime
    resolved_at: datetime | None
    conflict_summary: str | None
    grant_id: str | None
    deny_id: str | None
    remediation: list[str]
    context: dict


class ConflictListResponse(BaseModel):
    items: list[ConflictResponse]
    total: int
    page: int
    page_size: int


class UpdateConflictStatusRequest(BaseModel):
    status: Literal["investigating", "resolved", "dismissed"]
    resolution_note: str | None = Field(default=None, max_length=1024)


class ScanResult(BaseModel):
    conflicts_detected: int
    conflicts_created: int
    scan_duration_ms: int
    scanned_grants: int
    scanned_denies: int
    scanned_acl_mappings: int


# ── Explain-decision models ───────────────────────────────────────────────────


class ExplainDecisionRequest(BaseModel):
    subject_user_id: str
    resource_type: str
    resource_id: str | None = None
    action: str


class TraceStep(BaseModel):
    rule: str
    outcome: str  # "pass", "allow", "deny"
    detail: str | None = None


class ExplainDecisionResponse(BaseModel):
    decision: str  # "allow" or "deny"
    matched_rule: str
    deny_reason: str | None
    subject_user_id: str
    resource_type: str
    resource_id: str | None
    action: str
    trace: list[TraceStep]
    remediation: list[str]
    request_id: str
